status/views.py

```diff
@@ -1,97 +1,109 @@
+            connection_info.append({"name": "Mail Daemon", "status": mailusv_status, "details": mailusv_details})
+
+        return FormattedResponse(connection_info)
+
+    def test_redis_connection(self):
+        status, details = 'unknown', ''
+
+        try:
+            rs_details = settings.CONFIG['REDIS']
+            rs = redis.Redis(host=rs_details['HOST'], port=rs_details['PORT'])
+
+            if rs.ping() == True:
+                status = 'online'
             else:
-                data[2]['status'] = 'online'
-        else:
-            data[2]['status'] = 'unknown'
-            data[2]['details'] = 'Error connecting to /status'
-    except Exception as e:
-        data[2]['status'] = 'offline'
-        data[2]['details'] = str(e)
-
-    if settings.SEND_MAIL:
-        data.append(
-        {
-            "name": "Mail Daemon",
-            "status": "unknown",
-            "details": ''
-        })
+                status = 'offline'
+                details = "PING failed"
+        except redis.exceptions.RedisError as e:
+            status = 'offline'
+            details = "PING failed:\n" + str(e)
+
+        return status, details
+
+    def test_postgresql_connection(self):
+        status, details = 'unknown', ''
+
         try:
-            response = requests.get(settings.MAIL_SOCK_URL)
-            if response.status_code != 200:
-                data[3]['status'] = 'unknown'
-                data[3]['details'] = 'Unknown error whilst connecting to /'
+            with connection.cursor() as cursor:
+                cursor.execute("SELECT 1");
+            status = 'online'
+        except psycopg2.OperationalError as e:
+            status = 'offline'
+            details = 'Could not execute query `SELECT 1\':\n' + str(e)
+        
+        return status, details
+
+    def get_andromeda_status(self):
+        status, details = 'unknown', ''
+        try:
+            response = requests.get(f"{settings.CHALLENGE_SERVER_URL}/status",
+                                headers={"Authorization": settings.CHALLENGE_SERVER_API_KEY})
+            if response.status_code == 200:
+                sysinfo = response.json()
+                reasons = []
+
+                if sysinfo['Loads'][0] > 90:
+                    reasons.append(f"High CPU load: {sysinfo['Loads'][0]}")
+
+                if sysinfo['FreeRam'] / sysinfo['TotalRam'] < 0.2:
+                    reasons.append(f"High RAM usage: {sysinfo['FreeRam'] / 1024}MB free out of {sysinfo['TotalRam']}")
+
+                if len(reasons) != 0:
+                    status = 'partial'
+                    details = '\n'.join(reasons)
+                else:
+                    status = 'online'
             else:
-                data[3]['status'] = 'online'
-        except Exception as e:
-            data[3]['status'] = 'offline'
-            data[3]['details'] = str(e)
+                status = 'unknown'
```

Comment:
I'm not convinced about returning unknown, when you hit /status, one of a few things will happen, you'll get a 200, in which case, all good, the request is timed out or failed, in which case, its down, if it 500s, something really bad has happened, and it should be assumed to be down, if it 403s(might be a 401) then this should return something about an invalid api key

Reply:
Yes it was more a fallback if it wanted to get more complicated or we wanted to catch all generic exceptions so the route would always return something, but as I said earlier it's pretty redundant here.